Vungle/VungleRouter.m

```diff
@@ -503,59 +527,71 @@ - (BOOL)validateInfoData:(NSDictionary *)info
     @synchronized (self) {
-        if (placementID.length > 0) {
-            [self.delegatesDict removeObjectForKey:placementID];
-        } else if (state != BannerRouterDelegateStateUnknown) {
-            NSArray *array = [self.bannerDelegates.keyEnumerator allObjects];
-            for (NSString *key in array) {
-                if ([[self.bannerDelegates objectForKey:key] bannerState] == state) {
-                    [self.bannerDelegates removeObjectForKey:key];
-                }
+        NSArray *array = [self.bannerDelegates.keyEnumerator allObjects];
+        for (NSString *key in array) {
+            if ([[self.bannerDelegates objectForKey:key] bannerState] == state) {
+                [self.bannerDelegates removeObjectForKey:key];
+            }
+        }
+        array = [self.hbBannerDelegates.keyEnumerator allObjects];
+        for (NSString *key in array) {
+            if ([[self.hbBannerDelegates objectForKey:key] bannerState] == state) {
+                [self.hbBannerDelegates removeObjectForKey:key];
             }
         }
     }
 }
 
+- (void)cleanupFullScreenDelegate:(id<VungleRouterDelegate>)delegate
+{
+    @synchronized (self) {
+        NSString *placementID = [delegate getPlacementID];
+        NSString *eventID = [delegate getEventId];
+        if (eventID.length > 0) {
+            [self.hbDelegatesDict removeObjectForKey:eventID];
+        } else if (placementID.length > 0) {
+            [self.delegatesDict removeObjectForKey:placementID];
+        }
+    }
+}
+
+- (void)addToWaitingListWithDelegate:(id<VungleRouterDelegate>)delegate
+{
+    NSString *eventId = [delegate getEventId];
+    if (eventId.length > 0) {
+        [self.hbWaitingListDict setObject:delegate forKey:eventId];
+    } else {
+        NSString *placementId = [delegate getPlacementID];
+        if (![self.waitingListDict objectForKey:placementId]) {
+            [self.waitingListDict setObject:delegate forKey:placementId];
+        }
+    }
+}
+
 - (void)clearWaitingList
 {
-    for (id key in self.waitingListDict) {
-        id<VungleRouterDelegate> delegateInstance = [self.waitingListDict objectForKey:key];
+    [self requestAdsInWaitingListDictionary:self.waitingListDict];
+    [self requestAdsInWaitingListDictionary:self.hbWaitingListDict];
+}
 
+- (void)requestAdsInWaitingListDictionary:(NSMutableDictionary *)waitingListDict
+{
+    for (id key in waitingListDict) {
+        id<VungleRouterDelegate> delegateInstance = [waitingListDict objectForKey:key];
+        NSString *placementId = [delegateInstance getPlacementID];
+        
         if ([delegateInstance respondsToSelector:@selector(getBannerSize)]) {
-            NSString *id = [delegateInstance getPlacementID];
             CGSize size = [delegateInstance getBannerSize];
-            [self requestBannerAdWithPlacementID:id size:size delegate:delegateInstance];
+            [self requestBannerAdWithPlacementID:placementId size:size delegate:delegateInstance];
         } else {
```

Comment:
This `else` block is a copy and paste of the contents in `requestAdWithCustomEventInfo`